zarr/v3/store/local.py

```diff
@@ -0,0 +1,177 @@
+from __future__ import annotations
+
+import io
+import shutil
+from pathlib import Path
+from typing import Union, Optional, List, Tuple
+
+from zarr.v3.abc.store import Store
+from zarr.v3.common import BytesLike, concurrent_map, to_thread
+
+
+def _get(path: Path, byte_range: Optional[Tuple[int, Optional[int]]] = None) -> bytes:
+    if byte_range is not None:
+        start = byte_range[0]
+        end = (start + byte_range[1]) if byte_range[1] is not None else None
+    else:
+        return path.read_bytes()
+    with path.open("rb") as f:
+        size = f.seek(0, io.SEEK_END)
+        if start is not None:
+            if start >= 0:
+                f.seek(start)
+            else:
+                f.seek(max(0, size + start))
+        if end is not None:
+            if end < 0:
+                end = size + end
+            return f.read(end - f.tell())
+        return f.read()
+
+
+def _put(
+    path: Path,
+    value: BytesLike,
+    start: Optional[int] = None,
+    auto_mkdir: bool = True,
+):
+    if auto_mkdir:
+        path.parent.mkdir(parents=True, exist_ok=True)
+    if start is not None:
+        with path.open("r+b") as f:
+            f.seek(start)
+            f.write(value)
+    else:
+        return path.write_bytes(value)
+
+
+class LocalStore(Store):
+
+    supports_writes: bool = True
+    supports_partial_writes: bool = True
+    supports_listing: bool = True
+
+    root: Path
+    auto_mkdir: bool
+
+    def __init__(self, root: Union[Path, str], auto_mkdir: bool = True):
+        if isinstance(root, str):
+            root = Path(root)
+        assert isinstance(root, Path)
+
+        self.root = root
+        self.auto_mkdir = auto_mkdir
+
+    def __str__(self) -> str:
+        return f"file://{self.root}"
+
+    def __repr__(self) -> str:
+        return f"LocalStore({repr(str(self))})"
+
+    async def get(
+        self, key: str, byte_range: Optional[Tuple[int, Optional[int]]] = None
+    ) -> Optional[bytes]:
+        assert isinstance(key, str)
+        path = self.root / key
+
+        try:
+            return await to_thread(_get, path, byte_range)
+        except (FileNotFoundError, IsADirectoryError, NotADirectoryError):
+            return None
```

Comment:
I found this behavior in Zarrita to be surprising. @normanrz - can you comment on the motivation here? In V2 we would have just raised a KeyError.

Reply:
I opted for using `None` as non-existent. Raising and catching a KeyError would also work.